src/node/rpc/node_frontend.h

```diff
@@ -289,6 +289,35 @@ namespace ccf
             return make_success(rep);
           }
 
+          if (!this->context.get_node_state().is_primary())
```

Comment:
It seems to me that we could avoid all this code here if we created a new `endpoints::ForwardingRequired::RedirectOnly` enum that specified that the endpoint would never forward, but would redirect if the primary isn't us. The generic frontend could then handle this logic. Thoughts?

Reply:
Unless I'm missing something, it wouldn't help much because we still want to run much of the logic even if we're not primary, and in cases where we know the node is trusted, respond accordingly. The forwarding policy is accurately "NeverForward", if we were going to make this generic it might take the form of a wrapper, catching a RedirectException and returning the appropriate response.

I'm skeptical about going generic here though, this is a special case and we don't seem to have any other situation where this applicable?

Reply:
Fair enough. It's true that the read-only join can be served by any node, even if that node is retired. The code here looked at lot like the generic forwarding/redirect logic in `frontend.h`, hence my original comment.